in_toto/model.go

```diff
@@ -0,0 +1,188 @@
+  // I implemented it this way, because there will be several functions that
+  // receive or return a Metablock, where the inner signed has to be inferred
+  // on runtime, e.g. when iterating over links for a layout, and a link can
+  // turn out to be a layout (sublayout)
+  Signed interface{} `json:"signed"`
+  Signatures []Signature `json:"signatures"`
+}
+
+
+func (mb *Metablock) Load(path string) error {
+  // Open file and close before returning
+  jsonFile, err := os.Open(path)
+  defer jsonFile.Close()
+  if err != nil {
+    return err
+  }
+
+  // Read entire file
+  jsonBytes, err := ioutil.ReadAll(jsonFile)
+  if err != nil {
+    return err
+  }
+
+  // Unmarshal JSON into a map of raw messages (signed and signatures)
+  // We can't fully unmarshal immediately, because we need to inspect the
+  // type (link or layout) to decide which data structure to use
+  var rawMb map[string]*json.RawMessage
+  if err := json.Unmarshal(jsonBytes, &rawMb); err != nil {
+    return err
+  }
+
+  // Fully unmarshal signatures part
+  if err := json.Unmarshal(*rawMb["signatures"], &mb.Signatures); err != nil {
+    return err
+  }
+
+  // Temporarily copy signed to opaque map to inspect the `_type` of signed
+  // and create link or layout accordingly
+  var signed map[string]interface{}
+  if err := json.Unmarshal(*rawMb["signed"], &signed); err != nil {
+    return err
+  }
+
+  if signed["_type"] == "link" {
+    var link Link
+    if err := json.Unmarshal(*rawMb["signed"], &link); err != nil {
+    return err
+  }
+    mb.Signed = link
+
+  } else if signed["_type"] == "layout" {
+    var layout Layout
+    if err := json.Unmarshal(*rawMb["signed"], &layout); err != nil {
+    return err
+  }
+    mb.Signed = layout
+
+  } else {
+    return fmt.Errorf(`The '_type' field of the 'signed' part of a metadata
+        file must be one of 'link' or 'layout'`)
+  }
+
+  return nil
+}
+
+func (mb *Metablock) Dump(path string) error {
+  // JSON encode Metablock formatted with newlines and indentation
+  // TODO: parametrize format
+  jsonBytes, err := json.MarshalIndent(mb, "", "  ")
+  if err != nil {
+    return err
+  }
+
+  // Write JSON bytes to the passed path with permissions (-rw-r--r--)
+  err = ioutil.WriteFile(path, jsonBytes, 0644)
+  if err != nil {
+    return err
+  }
+
+  return nil
+}
+
+func (mb *Metablock) GetSignableRepresentation() ([]byte, error) {
+  return encode_canonical(mb.Signed)
+}
+
+
+func (mb *Metablock) VerifySignature(key Key) error {
+  var sig Signature
+  for _, s := range mb.Signatures {
+    if s.KeyId == key.KeyId {
+      sig = s
+      break
+    }
+  }
+
+  if sig == (Signature{}) {
+    return fmt.Errorf("No signature found for key '%s'", key.KeyId)
+  }
+
+  dataCanonical, err := mb.GetSignableRepresentation()
+  if err != nil {
+    return err
+  }
+
+  if err := VerifySignature(key, sig, dataCanonical); err != nil {
```

Comment:
Is that an assignment inside a condition, like the old C `if foo = bar` mishap? If so, I suggest to split into separate lines (assuming there are no side effects).

Also, below there seems to be another superfluous check for non-`nil` before returning.

Reply:
Re superfluous check for non-nil
See https://github.com/in-toto/in-toto-golang/pull/1#discussion_r230004116

Reply:
This is less often *used* in C I fear, it's mostly an *error* on behalf of the programmer :-)

I'm fine if this is considered sensible Go style.

Reply:
Same as answer to your comment above. What should the first value in the return tuple contain?

`VerifySignature` here is similar to the corresponding function in the Python reference implementation that returns nothing (on success) or raises an exception.

Re simplification:
Two against one, I surrender. :)
Is it okay if I replace expression-in-condition constructs everywhere in a separate PR?

Reply:
1, it's only 1.5 against 1 because this seems to be a Go idiom.

2, I'm fine with a separate PR.